### PR TITLE
Proper performant flex attention implementation

### PR DESCRIPTION
**Pull request by bursteratom**
# What does this PR do?

Current flex attention implementation does not take advantage of the performance and memory efficiency promised in [this official blog post from pytorch](https://pytorch.org/blog/flexattention/)

This PR, inspired by [https://github.com/pytorch/torchtune/blob/main/torchtune/modules/attention_utils.py](url) rectifies that by making flex attention always compiled and utilizing the sparse-optimised BlockMask data type for attention masking in lieu of regular torch tensor. Performance and memory utilization are now comparable to flash attention.

BlockMask creation has been implemented for the following models:

- [x] Llama

Let's add support for other models in a separate PR

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/integrations/flex_attention.py [hunk 1]
```diff
@@ -1,30 +1,117 @@
-from typing import Optional, Tuple
+from typing import Optional, Tuple, Union
 
 import torch
 
 from ..utils import is_torch_flex_attn_available
 
+"""
+Inspired by torchtune's flex attention implementation
+"""
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:9]: Nit: would move this to top of the file

File: src/transformers/integrations/flex_attention.py [hunk 2]
```diff
@@ -1,30 +1,117 @@
...
 if is_torch_flex_attn_available():
-    from torch.nn.attention.flex_attention import flex_attention
+    from torch.nn.attention.flex_attention import (
+        flex_attention,
+        BlockMask,
+        create_block_mask as create_block_causal_mask_flex,
+    )
+
+
+def make_flex_block_causal_mask(
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:19]: Nice!

File: src/transformers/integrations/flex_attention.py [hunk 3]
```diff
@@ -1,30 +1,117 @@
...
+    return create_block_causal_mask_flex(
+        causal_mask_mod,
+        batch_size,
+        None,
+        total_seq_len,
+        total_seq_len,
+        device=device,
+    )
+
+flex_attention_compiled = torch.compile(flex_attention, dynamic=False)
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:76]: This is unsafe and should rather be moved to at import time imo

File: src/transformers/integrations/flex_attention.py [hunk 4]
```diff
@@ -1,30 +1,117 @@
...
+        return causal_mask & document_mask & padding_mask
+
+    return create_block_causal_mask_flex(
+        causal_mask_mod,
+        batch_size,
+        None,
+        total_seq_len,
+        total_seq_len,
+        device=device,
+    )
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:74]: Would love kwargs in this case tbh

File: src/transformers/integrations/flex_attention.py [hunk 5]
```diff
@@ -33,11 +120,12 @@ def causal_mod(score, b, h, q_idx, kv_idx):
...
         return score
 
-    attn_output, attention_weights = flex_attention(
+    attn_output, attention_weights = compile_friendly_flex_attention(
         query,
         key,
         value,
-        score_mod=causal_mod,
+        score_mod=score_mod,
+        block_mask=block_mask,
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:128]: Is it necessary to pass both? What's the behavior if both is passed? Seems unnecessary.

Maybe move to the block mask option whenever possible and enable the score mod else. This is to have as much efficiency as possible (block mask) but a more powerful masking (score mod) when more unique attn types come up (e.g. softcapping).

**Review by vasqu — COMMENTED** [review]
A fan of flex attn :) hope you dont mind the comments but overall pro this. Torchtune is optimized for training iirc, is creating the block mask ok for inference? Like speed wise, I have no idea if there were any downsides/advantages for one or the other

Not sure if this is relevant to the PR tbh, but benchmarks might be a good thing to look out for in the future.

File: src/transformers/integrations/flex_attention.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:76]: Or even triggered later when I think about it. No need to compile things that might not even be used just because we have torch 2.5+

**Comment by vasqu** [comment]
Example for inference with flex attn: https://github.com/pytorch-labs/gpt-fast/pull/196

On first glance i can spot a few things:
- offset wrapper function
- create initial bigger sparse mask mod and index as necessary 
- compile block mask creation function

I think avoiding recreating the block mask is especially important here to avoid the memory/speed overhead - but not sure if I haven't measured speeds/memory myself. Might be more appropriate for a different PR, no idea; I just think inference especially should be handled with more care.

File: src/transformers/integrations/flex_attention.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/integrations/flex_attention.py:76]: Yep, and AFAIK flex gets compile automatically no?

File: src/transformers/integrations/flex_attention.py [hunk 6]
```diff
@@ -1,30 +1,117 @@
...
+
+    document_ids = attention_mask_2d
+    batch_size, total_seq_len = document_ids.shape
 
+    # Instead of passing a tensor mask, flex attention requires a mask_mod function
+    # that determines which elements of QK^T should be included in the attention
+    # computation prior to the softmax. For sample packing, we need both the
+    # logic for both causal mask and document mask. See PyTorch's official
+    # blog post for more details: https://pytorch.org/blog/flexattention/#mask-mods
+    def causal_mask_mod(b, h, q_idx, kv_idx):
```
Comment (ArthurZucker) [inline src/transformers/integrations/flex_attention.py:54]: let's just have better names, b is batch? h is hidden_state?

File: src/transformers/integrations/flex_attention.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/integrations/flex_attention.py:9]: Yep! And we forgot to add a licence!

File: src/transformers/models/llama/modeling_llama.py [hunk 7]
```diff
@@ -634,6 +634,10 @@ def _update_causal_mask(
             if attention_mask is not None and (attention_mask == 0.0).any():
                 return attention_mask
             return None
+        if self.config._attn_implementation == "flex_attention":
+            from ...integrations.flex_attention import make_flex_block_causal_mask
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:638]: to move at the top of the file with if is flex attention available!

File: src/transformers/integrations/flex_attention.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/integrations/flex_attention.py:128]: agree here! Score mod is used only because I was a bit too lazy to add the conversion of the mask, but we should only rely on block mask!

**Review by ArthurZucker — COMMENTED** [review]
Very much needed! Thanks a lot 🤗 
cc @molbap as you had issues with this recently !

File: src/transformers/integrations/flex_attention.py [hunk 8]
```diff
@@ -1,30 +1,117 @@
...
+        Defines the logic of a block causal mask by combining both a standard causal mask
+        and a block diagonal document mask.
+
+        See :func:`~torchtune.modules.attention_utils.create_block_causal_mask`
+        for an illustration.
+        """
+        causal_mask = q_idx >= kv_idx
+        document_mask = document_ids[b, q_idx] == document_ids[b, kv_idx]
+        padding_mask = document_ids[b, q_idx] > 0
+        return causal_mask & document_mask & padding_mask
```
Comment (molbap) [inline src/transformers/integrations/flex_attention.py:65]: To note, AFAIK the only thing here is that mask creation isn't compile-compatible because we can't (yet) index on several dimensions with `vmap` at compile time - for models with weird attention masks like `PaliGemma` it becomes a challenge, but it's overall fine as the attention itself is compiled!

**Review by molbap — COMMENTED** [review]
Big fan of this work! Thanks a lot for tackling it. I'd be interested in benchmarks especially in a couple models like PaliGemma and models with bidirectional attention :eyes:

File: src/transformers/integrations/flex_attention.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:76]: Nope, it doesn't sadly. Both should be compiled via `torch.compile` - the flex attn function `flex_attention ` itself as well as the block mask creation function `create_block_mask ` (not too sure about that one but should also be possible).

File: src/transformers/integrations/flex_attention.py (hunk 5 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:128]: I think in most cases block mask is possible but in certain other cases it is not, e.g.:
- softcapping
- t5 bias
- ...

Which means in some cases you need score mod for its expressiveness to manipulate scores directly imo.

File: src/transformers/integrations/flex_attention.py (hunk 6 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:54]: b(atch), h(ead), ... iirc

File: src/transformers/integrations/flex_attention.py (hunk 8 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:65]: I think you can somewhat compile with the `_compile` kwarg - at least it's a perf improvement.

File: src/transformers/integrations/flex_attention.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:76]: For reference: https://github.com/pytorch-labs/attention-gym/issues/95#issuecomment-2581607603

File: src/transformers/integrations/flex_attention.py (hunk 5 above)
Comment (bursteratom) [reply src/transformers/integrations/flex_attention.py:128]: yeah since score mod can do things like score modifying which mask mod can't do, I feel we should keep both. Use block mask for attention masking and score mod strictly for score manipulation. I can add a doc string about that

File: src/transformers/integrations/flex_attention.py (hunk 8 above)
Comment (bursteratom) [reply src/transformers/integrations/flex_attention.py:65]: yup blockmask creation was indeed uncompiled, and the whole thing still gives performance very similar to flash attention when tested on llama

Comment (molbap) [reply src/transformers/integrations/flex_attention.py:65]: amazing!

File: src/transformers/integrations/flex_attention.py (hunk 4 above)
Comment (bursteratom) [reply src/transformers/integrations/flex_attention.py:104]: Done!

File: src/transformers/integrations/flex_attention.py (hunk 3 above)
Comment (bursteratom) [reply src/transformers/integrations/flex_attention.py]: Thank you for your inputs @vasqu @ArthurZucker ! I wrapped the naive torch flex attention in a singleton class so that compilation happens only when triggered, and the compilation happens only once.

File: src/transformers/integrations/flex_attention.py (hunk 6 above)
Comment (bursteratom) [reply src/transformers/integrations/flex_attention.py]: yup, changed to a more descriptive variable name

File: src/transformers/integrations/flex_attention.py (hunk 1 above)
Comment (bursteratom) [reply src/transformers/integrations/flex_attention.py]: @ArthurZucker can you point me to an example of how a proper licence string should be added?

**Comment by bursteratom** [comment]
@vasqu @molbap @ArthurZucker I made some changes according to your inputs, wondering if you can give it another pass? Thank you!

File: src/transformers/integrations/flex_attention.py [hunk 9]
```diff
@@ -1,43 +1,161 @@
...
 if is_torch_flex_attn_available():
-    from torch.nn.attention.flex_attention import flex_attention
+    from torch.nn.attention.flex_attention import (
+        flex_attention,
+        BlockMask,
+        create_block_mask as create_block_causal_mask_flex,
+    )
+
+
+class wrapped_flex_attention:
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:20]: More of a nit but class names should be pascal case imo

File: src/transformers/integrations/flex_attention.py [hunk 10]
```diff
@@ -1,43 +1,161 @@
...
+    """
+    We are doing a singleton class so that flex attention is compiled once and only once
+    and only when called.
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:23]:
```suggestion
    We are doing a singleton class so that flex attention is compiled once when it's first called.
```

File: src/transformers/integrations/flex_attention.py [hunk 11]
```diff
@@ -1,43 +1,161 @@
...
+        return causal_mask & document_mask & padding_mask
+
+    return create_block_causal_mask_flex(
+        causal_mask_mod,
+        batch_size,
+        None,
+        Q_LEN=total_seq_len,
+        KV_LEN=total_seq_len,
+        device=device,
+    )
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:104]: I think my marking last time made it a bit confusing - kwargs on all args would be beneficial imo, especially on the `None` arg (attention heads).

File: src/transformers/integrations/flex_attention.py [hunk 12]
```diff
@@ -1,43 +1,161 @@
...
+
+@torch.compiler.disable(recursive=False)
+def compile_friendly_flex_attention(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    **kwargs,
+) -> torch.Tensor:
+
+    flex_attention_compiled = wrapped_flex_attention()()
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:114]:
```suggestion
    flex_attention_compiled = wrapped_flex_attention()
```
Honestly not sure, guess it's a typo

File: src/transformers/integrations/flex_attention.py (hunk 5 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:158]: Still think it would be nice to be more explicit here and cast `score_mod=None` if we don't need it. Don't even let the implementation think about it.

**Review by vasqu — APPROVED** [review]
Honestly, think the core is fine - just a few nits and smaller things. Would leave inference for another PR :)

File: src/transformers/integrations/flex_attention.py (hunk 12 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:113]: Ah ic, that makes sense. More of a lack of understanding on my side.

Would at least add a comment on what's happening, in isolation it does look weird at first glance.

Comment (bursteratom) [reply src/transformers/integrations/flex_attention.py:113]: Will do! Thank you very much for your feedback!

**Review by ArthurZucker — COMMENTED** [review]
Very nice! 
Just missing some doc/ small perf comparisons!

File: src/transformers/integrations/flex_attention.py (hunk 5 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:160]: I meant it more like:
- 1. If we don't need, i.e. simple causal inference (with padding) -> block mask as it's the most efficient flex attn implementation 
- 2. If we have a head mask and/or other special score modifications (softcap) -> score mod as block mask is not able to modify the scores like we want to

But only allow one to be passed depending on the usecase. My assumption is that torch might call both and slow down the forward (due to the unnecessary score mod) but that should be checked. Does that make sense?

Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:160]: Ok after looking into the source code ( https://github.com/pytorch/pytorch/blob/6ca5c22e31635df6a8b8b9575e21d769fbeb178c/torch/nn/attention/flex_attention.py#L1260 ), it seems score mods are always given - i.e. it doesn't make a difference passing the score mod (esp since you only modify scores if softacpping or head masking). So feel free to close this! It should be fine the way it is.

File: src/transformers/integrations/flex_attention.py (hunk 9 above)
Comment (bursteratom) [reply src/transformers/integrations/flex_attention.py]: fixed!

File: src/transformers/integrations/flex_attention.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py]: I think something along these lines is meant https://github.com/huggingface/transformers/blob/d18d9c3205323d02479ca0d7cfd978908ebb6f71/src/transformers/models/siglip2/modular_siglip2.py#L1-L14

(you can add the torchtune or pytorch team imo, not sure how finegrained it should be)

**Review by ArthurZucker — APPROVED** [review]
LGTM! could you just add some documentation about perffs comparison ! 🤗

**Review by shethaadit — APPROVED** [review]

**Comment by bursteratom** [comment]
@ArthurZucker thank you! I will add the doc and perf comparison shortly!

**Comment by ArthurZucker** [comment]
Let's merge for now IMO and you can open a new PR for doc!

**Comment by ArthurZucker** [comment]
We can close PR is merged! 🤗
